File: Moxion/Domain/Units/KinematicUnitInfo.cs
using Moxion.Abstractions;
using Moxion.Common.Units;

namespace Moxion.Domain.Units;

public readonly record struct KinematicUnitInfo(
  PositionUnitInfo Position,
  VelocityUnitInfo Velocity,
  AccelerationUnitInfo Acceleration,
  JerkUnitInfo Jerk,
  TimeUnitInfo Time
) : IUnitInfo<KinematicUnitInfo>
{
  public static KinematicUnitInfo Create( PositionUnit positionUnit, TimeUnit timeUnit )
  {
    return new KinematicUnitInfo(
      new PositionUnitInfo( positionUnit ),
      new VelocityUnitInfo( positionUnit, timeUnit ),
      new AccelerationUnitInfo( positionUnit, timeUnit ),
      new JerkUnitInfo( positionUnit, timeUnit ),
      new TimeUnitInfo( timeUnit )
    );
  }
}

File: Moxion.Presentation.Web/ClientApp/src/models/kinematics/units/KinematicUnitInfo.ts
import PositionUnitInfo from "./PositionUnitInfo";
import VelocityUnitInfo from "./VelocityUnitInfo";
import JerkUnitInfo from "./JerkUnitInfo";
import AccelerationUnitInfo from "./AccelerationUnitInfo";

class KinematicUnitInfo {
    displacement: PositionUnitInfo
    velocity: VelocityUnitInfo
    acceleration: AccelerationUnitInfo
    jerk: JerkUnitInfo

    constructor(
        displacement?: PositionUnitInfo,
        velocity?: VelocityUnitInfo,
        acceleration?: AccelerationUnitInfo,
        jerk?: JerkUnitInfo
    ) {
        this.displacement = displacement ?? new PositionUnitInfo('None');
        this.velocity = velocity ?? new VelocityUnitInfo('None', 'None');
        this.acceleration = acceleration ?? new AccelerationUnitInfo('None', 'None');
        this.jerk = jerk ?? new JerkUnitInfo('None', 'None');
    }

    static from(other: KinematicUnitInfo | null) {
        if (!other) {
            return new KinematicUnitInfo()
        }

        return new KinematicUnitInfo(
            PositionUnitInfo.from(other.displacement),
            VelocityUnitInfo.from(other.velocity),
            AccelerationUnitInfo.from(other.acceleration),
            JerkUnitInfo.from(other.jerk),
        )
    }
}

export default KinematicUnitInfo